### PR TITLE
Upgrade to LLVM 6 and use BinaryBuilder for LLVM on CI

### PR DESCRIPTION
**Pull request by vchuravy**
This is a work-in-progress PR for #26398.
This uses the BinaryBuilder infrastructure to create tarballs
with our patchset, that we can use to seed the Mac OSX and Windows
CIs. See https://github.com/staticfloat/LLVMBuilder for the buildscripts.

Many thanks to @staticfloat

### CODE REVIEW

File: deps/llvm.mk [hunk 1]
```diff
@@ -566,3 +567,35 @@ ifeq ($(USE_POLLY),1)
...
+	cp $< $@
+
+clean-llvm:
+distclean-llvm:
+get-llvm:
+extract-llvm:
+configure-llvm:
+compile-llvm:
+fastcheck-llvm:
+check-llvm:
```
Comment (vtjnash) [inline deps/llvm.mk]: need to define whichever of these are applicable (probably just `get`?)

**Comment by vchuravy** [comment]
Intermittent status report:

The build-system changes seem to work, but we hit issues with the prepared tarballs.

- https://github.com/staticfloat/LLVMBuilder/issues/5
- https://github.com/staticfloat/LLVMBuilder/issues/6

**Comment by vchuravy** [comment]
Good news!

All test pass. CI is red because apparently httpbin.org got hit by a network fluke and that caused one test to fail across Circle, Travis, and Appveyor. 

At this point I would encourage testing and reviews by others and I would like to see this merged soon (tm).

File: Make.inc [hunk 2]
```diff
@@ -220,6 +220,11 @@ includedir_rel := $(shell $(JULIAHOME)/contrib/relative_path.sh $(bindir) $(incl
 INSTALL_F := $(JULIAHOME)/contrib/install.sh 644
 INSTALL_M := $(JULIAHOME)/contrib/install.sh 755
 
+# BinaryBuilder options
+# TODO: Autodiscover triplet
```
Comment (ararslan) [inline Make.inc]: I think `$(shell $(CC) -dumpmachine)` should work in the majority of cases, with two exceptions: cross-compiling (in which case we already know the target triplet, right?) and platforms such as macOS and FreeBSD, where the minimum version number we support is part of the triplet. But if BinaryBuilder does triplet mapping using the `Platform` type, the latter case shouldn't be an issue, I would think.

File: Make.inc [hunk 3]
```diff
@@ -220,6 +220,11 @@ includedir_rel := $(shell $(JULIAHOME)/contrib/relative_path.sh $(bindir) $(incl
...
+USE_BINARYBUILDER_LLVM := 0
+BB_TRIPLET :=
```
Comment (ararslan) [inline Make.inc]: If this is intended to be at all user-facing, could we use a more descriptive name, perhaps just spelling out BinaryBuilder as in the `USE_BINARYBUILDER_LLVM` variable above?

File: deps/llvm.mk [hunk 4]
```diff
@@ -566,3 +567,35 @@ ifeq ($(USE_POLLY),1)
...
+	$(JLDOWNLOAD) $@ $(LLVM_BB_URL)
+
+$(BUILDDIR)/llvm-$(LLVM_VER)-$(LLVM_BB_REL)/build-compiled: | $(BUILDDIR)/llvm-$(LLVM_VER)-$(LLVM_BB_REL)/LLVM.$(BB_TRIPLET).tar.gz
+	echo 1 > $@
+
+$(eval $(call staged-install,llvm,llvm-$$(LLVM_VER)-$$(LLVM_BB_REL),,,,))
+
+#Override provision of stage tarball
+$(build_staging)/llvm-$(LLVM_VER)-$(LLVM_BB_REL).tgz: $(BUILDDIR)/llvm-$(LLVM_VER)-$(LLVM_BB_REL)/build-compiled
+	$(info Using staged install form staticfloat/LLVMBuilder)
```
Comment (ararslan) [inline deps/llvm.mk]: form -> from?

File: src/codegen.cpp [hunk 5]
```diff
@@ -5493,9 +5493,7 @@ static std::unique_ptr<Module> emit_function(
     AttrBuilder *attr = new AttrBuilder();
     attr->addStackAlignmentAttr(16);
 #if JL_LLVM_VERSION >= 50000
-    f->addAttributes(AttributeList::FunctionIndex,
-        AttributeList::get(f->getContext(),
-            AttributeList::FunctionIndex, *attr));
```
Comment (ararslan) [inline src/codegen.cpp]: Is this not still applicable to LLVM 5.0?

Comment (vchuravy) [reply src/codegen.cpp]: I think it was just never tested.

File: Make.inc (hunk 2 above)
Comment (vchuravy) [reply Make.inc]: Yeah it is very close, but not quite and I don't want to spend time on figuring out the precise differences. So for now this has to be a conscious choice.

**Comment by vchuravy** [comment]
Wohaaa straight flush.

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/26925#issuecomment-385674419) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/84b631769bd3da2be01f47be6c2f05043e2a0beb/b5199af_vs_132a9f4/report.md). cc @ararslan

**Comment by vchuravy** [comment]
I will merge this on Friday if there aren't any objections or comments that require big reworks.

**Comment by ararslan** [comment]
We don't have CI set up for Alpine Linux so I'm building and running tests for this branch on Alpine in a Docker container now. I'll report back with my findings.

**Comment by ararslan** [comment]
The situation on Alpine doesn't change with this PR; all tests pass except for those noted in #26761 (`cmdlineargs` and `precompile`).

**Comment by fredrikekre** [comment]
Seems like all benchmarks with `BigInt` is ~1.5x-2x times slower, whats up with that?

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(ALL, vs = "509d6a1a8871dcec78cd6603b6d64871d1088b2f")`

**Comment by KristofferC** [comment]
Even if the BigInt benchmarks happen to be a bit slower here, I don't think that should block this PR.

**Comment by ararslan** [comment]
Restarted Nanosoldier. Looks like your invocation had a syntax error though: specific SHAs need to be prefixed with `@`.

@nanosoldier `runbenchmarks(ALL, vs="@509d6a1a8871dcec78cd6603b6d64871d1088b2f")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/26925#issuecomment-386063885) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/f8e70e8693ddc79feee81375c1013ceed3e26771/fb09d28_vs_509d6a1/report.md). cc @ararslan

**Comment by Keno** [comment]
Yeah, you'll want to rebase.

**Comment by vchuravy** [comment]
master at fa0704b8b8c2554d91249f8d1636a59b93ac14e2 only triggered with new optimizer.

```
     From worker 7:	julia: /scratch/vchuravy/julia/usr/include/llvm/Support/Casting.h:255: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = llvm::PointerType; Y = llvm::Type; typename llvm::cast_retty<X, Y*>::ret_type = llvm::PointerType*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
      From worker 7:	
      From worker 7:	signal (6): Aborted
      From worker 7:	in expression starting at /scratch/vchuravy/julia/test/strings/search.jl:9
      From worker 7:	raise at /build/glibc-Cl5G7W/glibc-2.23/signal/../sysdeps/unix/sysv/linux/raise.c:54
      From worker 7:	abort at /build/glibc-Cl5G7W/glibc-2.23/stdlib/abort.c:89
      From worker 7:	__assert_fail_base at /build/glibc-Cl5G7W/glibc-2.23/assert/assert.c:92
      From worker 7:	__assert_fail at /build/glibc-Cl5G7W/glibc-2.23/assert/assert.c:101
      From worker 7:	cast<llvm::PointerType, llvm::Type> at /scratch/vchuravy/julia/usr/include/llvm/Support/Casting.h:255
      From worker 7:	getValueAddrSpace at /scratch/vchuravy/julia/src/llvm-late-gc-lowering.cpp:446 [inlined]
      From worker 7:	NumberBase at /scratch/vchuravy/julia/src/llvm-late-gc-lowering.cpp:589
      From worker 7:	Number at /scratch/vchuravy/julia/src/llvm-late-gc-lowering.cpp:612
      From worker 7:	LocalScan at /scratch/vchuravy/julia/src/llvm-late-gc-lowering.cpp:1156
      From worker 7:	runOnFunction at /scratch/vchuravy/julia/src/llvm-late-gc-lowering.cpp:2113
      From worker 7:	runOnFunction at /scratch/vchuravy/julia/deps/srccache/llvm-6.0.0/lib/IR/LegacyPassManager.cpp:1520
      From worker 7:	runOnModule at /scratch/vchuravy/julia/deps/srccache/llvm-6.0.0/lib/IR/LegacyPassManager.cpp:1541
      From worker 7:	run at /scratch/vchuravy/julia/deps/srccache/llvm-6.0.0/lib/IR/LegacyPassManager.cpp:1597
      From worker 7:	operator() at /scratch/vchuravy/julia/src/jitlayers.cpp:466
      From worker 7:	addModule at /scratch/vchuravy/julia/usr/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h:57 [inlined]
      From worker 7:	addModule at /scratch/vchuravy/julia/src/jitlayers.cpp:593
      From worker 7:	jl_add_to_ee at /scratch/vchuravy/julia/src/jitlayers.cpp:831 [inlined]
      From worker 7:	jl_finalize_function at /scratch/vchuravy/julia/src/jitlayers.cpp:839
      From worker 7:	getAddressForFunction at /scratch/vchuravy/julia/src/codegen.cpp:1317
      From worker 7:	jl_generate_fptr at /scratch/vchuravy/julia/src/codegen.cpp:1428
      From worker 7:	jl_compile_method_internal at /scratch/vchuravy/julia/src/gf.c:1830 [inlined]
      From worker 7:	jl_fptr_trampoline at /scratch/vchuravy/julia/src/gf.c:1837
      From worker 7:	jl_toplevel_eval_flex at /scratch/vchuravy/julia/src/toplevel.c:850
      From worker 7:	jl_parse_eval_all at /scratch/vchuravy/julia/src/ast.c:852
      From worker 7:	jl_load at /scratch/vchuravy/julia/src/toplevel.c:894
      From worker 7:	include at ./boot.jl:314 [inlined]
      From worker 7:	include_relative at ./loading.jl:1067
      From worker 7:	include at ./sysimg.jl:29
      From worker 7:	include at /scratch/vchuravy/julia/test/testdefs.jl:13
      From worker 7:	jl_fptr_trampoline at /scratch/vchuravy/julia/src/gf.c:1838
      From worker 7:	jl_apply_generic at /scratch/vchuravy/julia/src/gf.c:2167
      From worker 7:	macro expansion at /scratch/vchuravy/julia/test/testdefs.jl:22 [inlined]
      From worker 7:	macro expansion at /scratch/vchuravy/julia/usr/share/julia/stdlib/v0.7/Test/src/Test.jl:1079 [inlined]
      From worker 7:	macro expansion at /scratch/vchuravy/julia/test/testdefs.jl:21 [inlined]
      From worker 7:	macro expansion at ./util.jl:289 [inlined]
      From worker 7:	top-level scope at ./<missing>:19
```

**Comment by Keno** [comment]
I'll take a look at that failure.

**Comment by ararslan** [comment]
Awesome work here, Valentin!

**Pull request merged**